File: src/main/java/upbrella/be/store/entity/StoreDetail.java
package upbrella.be.store.entity;

import lombok.*;
import upbrella.be.store.dto.request.CreateStoreRequest;
import upbrella.be.store.dto.request.UpdateStoreRequest;

import javax.persistence.*;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

@Entity
@Getter
@Builder
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@AllArgsConstructor
public class StoreDetail {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private long id;
    @OneToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "store_meta_id")
    private StoreMeta storeMeta;
    private String umbrellaLocation;
    private String workingHour;
    private String instaUrl;
    private String contactInfo;
    private String address;
    private String addressDetail;
    private String content;
    @OneToMany(mappedBy = "storeDetail", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    private List<StoreImage> storeImages;


    public static StoreDetail createForSave(CreateStoreRequest request, StoreMeta storeMeta) {

        return StoreDetail.builder()
                .storeMeta(storeMeta)
                .umbrellaLocation(request.getUmbrellaLocation())
                .workingHour(request.getBusinessHour())
                .instaUrl(request.getInstagramId())
                .contactInfo(request.getContactNumber())
                .address(request.getAddress())
                .addressDetail(request.getAddressDetail())
                .content(request.getContent())
                .build();
    }

    public void updateStore(StoreMeta storeMeta, UpdateStoreRequest request) {

        this.storeMeta = storeMeta;
        this.umbrellaLocation = request.getUmbrellaLocation();
        this.workingHour = request.getBusinessHour();
        this.instaUrl = request.getInstagramId();
        this.contactInfo = request.getContactNumber();
        this.address = request.getAddress();
        this.addressDetail = request.getAddressDetail();
        this.content = request.getContent();
    }

    public List<StoreImage> getSortedStoreImages() {

        return storeImages.stream()
                .sorted((o1, o2) -> (int) (o1.getId() - o2.getId()))
                .collect(Collectors.toList());
    }
}

File: src/main/kotlin/upbrella/be/store/entity/StoreDetail.kt
package upbrella.be.store.entity

import upbrella.be.store.dto.request.CreateStoreRequest
import upbrella.be.store.dto.request.UpdateStoreRequest
import javax.persistence.*

@Entity
class StoreDetail(
    @OneToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "store_meta_id")
    var storeMeta: StoreMeta? = null,
    var umbrellaLocation: String? = null,
    var workingHour: String? = null,
    var instaUrl: String? = null,
    var contactInfo: String? = null,
    var address: String? = null,
    var addressDetail: String? = null,
    var content: String? = null,
    @OneToMany(mappedBy= "storeDetail", fetch = FetchType.LAZY, cascade = [CascadeType.ALL])
    val storeImages: List<StoreImage> = emptyList(),
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long? = null,
) {
    companion object {
        @JvmStatic
        fun createForSave(request: CreateStoreRequest, storeMeta: StoreMeta): StoreDetail {
            return StoreDetail(
                storeMeta = storeMeta,
                umbrellaLocation = request.umbrellaLocation,
                workingHour = request.businessHour,
                instaUrl = request.instagramId,
                contactInfo = request.contactNumber,
                address = request.address,
                addressDetail = request.addressDetail,
                content = request.content,
            )
        }
    }

    fun updateStore(storeMeta: StoreMeta, request: UpdateStoreRequest) {
        this.storeMeta = storeMeta
        this.umbrellaLocation = request.umbrellaLocation
        this.workingHour = request.businessHour
        this.instaUrl = request.instagramId
        this.contactInfo = request.contactNumber
        this.address = request.address
        this.addressDetail = request.addressDetail
        this.content = request.content
    }

    fun getSortedStoreImages(): List<StoreImage> {
        return storeImages.stream()
            .sorted { o1, o2 -> (o1.id - o2.id).toInt() }
            .toList()
    }
}